registry/storage/revisionstore.go

```diff
@@ -131,16 +160,25 @@ func (rs *revisionStore) link(revision digest.Digest) error {
 	return rs.blobStore.link(revisionPath, revision)
 }
 
-// delete removes the specified manifest revision from storage.
+// delete creates a tombstone for the given digest revision.  The tombstone
+// is a link file which points to the deleted manifest.
 func (rs *revisionStore) delete(revision digest.Digest) error {
-	revisionPath, err := rs.pm.path(manifestRevisionPathSpec{
-		name:     rs.Name(),
-		revision: revision,
-	})
-
+	// don't allow deleting a blob that doesn't already exist
+	exists, err := rs.exists(revision)
 	if err != nil {
 		return err
 	}
 
-	return rs.driver.Delete(rs.repository.ctx, revisionPath)
+	if !exists {
+		return distribution.ErrUnknownManifestRevision{
+			Name:     rs.Name(),
+			Revision: revision,
+		}
+	}
+
+	if err := rs.tomb.putTombstone(rs.repository.ctx, rs.Name(), revision); err != nil {
+		return err
+	}
+
```

Comment:
@mountkin This is a consistency issue. The output of `/tag/list` can likely never be consistent without making the call extremely inefficient. One cannot simply delete the tag for current, since another manifest may have that tag. For example, if there are three (X, Y, and Z) manifests tagged A, and A is currently pointed to Z, since it was pushed last. If you delete Z, where should A point? If you say Y, the registry needs to keep track of the order in which the tags were pushed. If you say the reference should disappear, can X and Y no longer be referenced by the tag?